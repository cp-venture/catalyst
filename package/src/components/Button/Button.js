import React from "react";
import PropTypes from "prop-types";
import CircularProgress from "@material-ui/core/CircularProgress";
import MuiButton from "@material-ui/core/Button";
import makeStyles from "@material-ui/core/styles/makeStyles";

const useStyles = makeStyles((theme) => ({
  buttonProgress: {
    marginLeft: theme.spacing.unit
  },
  containedPrimary: {
    "color": theme.palette.primary.contrastText,
    "backgroundColor": theme.palette.colors.red,
    "&:hover": {
      "backgroundColor": theme.palette.colors.redHover,
      // Reset on touch devices, it doesn't add specificity
      "@media (hover: none)": {
        backgroundColor: theme.palette.colors.redHover
      }
    }
  },
  outlinedPrimary: {
    "color": theme.palette.colors.red,
    "border": `1px solid ${theme.palette.colors.red}`,
    "&:hover": {
      "border": `1px solid ${theme.palette.colors.redBorder}`,
      "backgroundColor": theme.palette.colors.redBackground,
      // Reset on touch devices, it doesn't add specificity
      "@media (hover: none)": {
        backgroundColor: "transparent"
      }
    }
  }
}));
/**
 * @name Button
 * @param {Object} props Component props
 * @returns {React.Component} returns a React component
 */
function Button(props) {
  const {
    children,
    color,
    disabled,
    isWaiting,
    ...otherProps
  } = props;

  const classes = useStyles();

  if (color === "danger") {
    return (
      <MuiButton
        classes={{
          containedPrimary: classes.containedPrimary,
          outlinedPrimary: classes.outlinedPrimary
        }}
        color="primary"
        disabled={disabled || isWaiting}
        {...otherProps}
      >
        {children}
        {isWaiting && <CircularProgress size={16} className={classes.buttonProgress} />}
      </MuiButton>
    );
  }

  return (
    <MuiButton
      color={color}
      disabled={disabled || isWaiting}
      {...otherProps}
    >
      {children}
      {isWaiting && <CircularProgress size={16} className={classes.buttonProgress} />}
    </MuiButton>
  );
}

Button.propTypes = {
  /**
   * What goes here?
   */
  children: PropTypes.node,
  /**
   * CSS Classes
   */
  classes: PropTypes.object,
  /**
   * Options: `default` | `inherit` | `primary` | `secondary` | `danger`
   */
  color: PropTypes.enum,
  /**
   * If `true`, the button will be disabled.
   */
  disabled: PropTypes.bool, // eslint-disable-line
  /**
   * If `true`, the CircularProgress will be displayed.
   */
  isWaiting: PropTypes.bool,
  /**
   * onClick callback
   */
  onClick: PropTypes.func
};

export default Button;
