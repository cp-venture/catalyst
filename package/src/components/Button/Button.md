### Overview

The Catalyst Button inherits from the Material-UI [Button component](https://material-ui.com/components/buttons/). Refer to tne Material-UI [Button API docs](https://material-ui.com/api/button/) for more information. 

### Usage


#### Default Catalyst button

This is what a button with all the default prop options looks like:

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button>Text-only</Button>
  </div>
</div>
```

It's a button with `variant` set to `text`, `size` set to `medium`, `color` set to `default`.

#### Material UI options

- **Solid button**: The solid button is used for a primary action in a modal, card, large view and generally throughout.
- **Outline button**: The outline button is used for a secondary or dismissive action. The outline button should be paired with the solid button in cases such as dismissing a modal or canceling an action.

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button>Text-only</Button>
  </div>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="contained">Solid</Button>
  </div>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="outlined">Outlined</Button>
  </div>
</div>
```

- **Disabled**: Docs go here. When do you use this button in Catalyst?

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="contained" isWaiting>isWaiting</Button>
  </div>
</div>
```

- **Full Width**: Docs go here. When do you use this button in Catalyst?

```jsx
<div style={{ display: "block" }}>
  <Button variant="contained" fullWidth>FullWidth</Button>
</div>
```

#### Catalyst-custom buttons

- **Danger button**: The danger button is used for a destructive action that is difficult to recover from such as deleting information. The danger button should be used at the point that the destructive action actually takes place. For example, you can have a delete button as a secondary action on a page and in this case you would use a an outline button, the outline button would then trigger a modal confirmation, which is where you would use the danger button.

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="contained" color="danger">Danger - Contained</Button>
  </div>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="outlined" color="danger">Danger - Outlined</Button>
  </div>
</div>
```

- **isWaiting**: Docs go here.

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="contained" isWaiting>isWaiting - Contained</Button>
  </div>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="outlined" isWaiting>isWaiting - Outlined</Button>
  </div>
</div>
```

