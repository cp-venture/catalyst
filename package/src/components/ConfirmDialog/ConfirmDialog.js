import React, { Fragment, useState } from "react";
import PropTypes from "prop-types";
import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogContentText from "@material-ui/core/DialogContentText";
import DialogTitle from "@material-ui/core/DialogTitle";

/**
 * @name ConfirmDialog
 * @param {Object} props Component props
 * @returns {React.Component} A React component
 */
function ConfirmDialog(props) {
  const { children, title, message, confirmActionText, cancelActionText, onConfirm } = props;
  const [isOpen, setIsOpen] = useState(false);

  const handleClose = () => {
    setIsOpen(false);
  };

  return (
    <Fragment>
      {children({
        openDialog: () => {
          setIsOpen(true);
        }
      })}
      <Dialog
        aria-labelledby="confirm-action-dialog-title"
        maxWidth="sm"
        onClose={handleClose}
        open={isOpen}
      >
        <DialogTitle id="confirm-action-dialog-title">{title}</DialogTitle>
        {message && (
          <DialogContent>
            <DialogContentText>{message}</DialogContentText>
          </DialogContent>
        )}

        <DialogActions>
          <Button onClick={handleClose} color="primary" variant="outlined">
            {cancelActionText}
          </Button>
          <Button
            onClick={() => {
              onConfirm();
              setIsOpen(false);
            }}
            color="primary"
            variant="contained"
          >
            {confirmActionText}
          </Button>
        </DialogActions>
      </Dialog>
    </Fragment>
  );
}


ConfirmDialog.propTypes = {
  /**
   * Cancel button text
   */
  cancelActionText: PropTypes.string,
  /**
   * Render prop `{({ openDialog }) => ()}`
   */
  children: PropTypes.func.isRequired,
  /**
   * Text for confirm button
   */
  confirmActionText: PropTypes.string,
  /**
   * Message body. May be a string or a React component.
   */
  message: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  /**
   * Confirmation callback
   */
  onConfirm: PropTypes.func,
  /**
   * Dialog title
   */
  title: PropTypes.oneOfType([PropTypes.string, PropTypes.node])
};

ConfirmDialog.defaultProps = {
  cancelActionText: "Cancel",
  confirmActionText: "OK",
  onConfirm() { }
};

export default ConfirmDialog;
