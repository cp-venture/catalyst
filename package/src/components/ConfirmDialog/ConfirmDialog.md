### Overview

The `ConfirmDialog` component uses [Dialog](https://material-ui.com/api/dialog), [DialogActions](https://material-ui.com/api/dialog-actions/), [DialogContent](https://material-ui.com/api/dialog-content/), [DialogContentText](https://material-ui.com/api/dialog-content-text/), [Button](https://material-ui.com/api/button/) from Material-UI and [DialogTitle](http://localhost:4060/#/Base%20Components/Feedback/DialogTitle) from Catalyst.

### Usage

The `ConfirmDialog` component provides a standard confirmation alert base on some user action. Pass a callback function to `onConfirm()` to allow the user to take an action after clicking the confirmation button.

```jsx
import Button from "../Button";

<ConfirmDialog
  title="Archive 24 products?"
  message="Archiving products removes them from both admin and customer views."
  onConfirm={() => { alert("Action confirmed!") }}
>
  {({ openDialog }) => (
    <Button color="primary" onClick={openDialog} variant="contained">Open Confirm Dialog</Button>
  )}
</ConfirmDialog>
```
