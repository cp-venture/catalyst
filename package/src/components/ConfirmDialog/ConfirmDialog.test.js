import React from "react";
import { render, fireEvent, cleanup } from "../../tests/";
import "@testing-library/jest-dom/extend-expect";
import Button from "../Button";
import ConfirmDialog from "./ConfirmDialog";

afterEach(cleanup);

test("basic snapshot - with opening the dialog", () => {
  /* eslint-disable function-paren-newline */
  const { asFragment, getByText, getByRole } = render(
    <ConfirmDialog
      title="Are you sure?"
      message="Are you sure you want to do that?"
    >
      {({ openDialog }) => (
        <Button color="primary" onClick={openDialog} variant="contained">Open Confirm Dialog</Button>
      )}
    </ConfirmDialog>);
  fireEvent.click(getByText("Open Confirm Dialog"));
  expect(getByRole("dialog")).toHaveTextContent("Are you sure you want to do that?");
  expect(getByRole("dialog")).toHaveTextContent("Are you sure?");
  expect(getByRole("dialog")).toHaveTextContent("OK");
  expect(getByRole("dialog")).toHaveTextContent("Cancel");
  expect(getByRole("dialog")).toMatchSnapshot();
  expect(asFragment()).toMatchSnapshot();
});
