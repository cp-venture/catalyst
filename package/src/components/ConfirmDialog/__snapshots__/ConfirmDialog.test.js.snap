// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot - with opening the dialog 1`] = `
<div
  aria-labelledby="confirm-action-dialog-title"
  class="MuiDialog-root"
  role="dialog"
  style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiDialog-container MuiDialog-scrollPaper"
    role="document"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    tabindex="-1"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth MuiPaper-rounded"
    >
      <div
        class="MuiDialogTitle-root"
        id="confirm-action-dialog-title"
      >
        <h2
          class="MuiTypography-root MuiTypography-h4"
        >
          Are you sure?
        </h2>
      </div>
      <div
        class="MuiDialogContent-root"
      >
        <p
          class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorInherit"
        >
          Are you sure you want to do that?
        </p>
      </div>
      <div
        class="MuiDialogActions-root MuiDialogActions-spacing"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Cancel
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            OK
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;

exports[`basic snapshot - with opening the dialog 2`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Open Confirm Dialog
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
